waspc/src/Wasp/Generator/NpmInstall.hs

```diff
@@ -101,25 +102,27 @@ loadInstalledFullStackNpmDependencies dstDir = do
 installNpmDependencies :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
 installNpmDependencies projectDir = do
   chan <- newChan
-  let runSetupJobs = concurrently (ServerSetup.installNpmDependencies projectDir chan) (WebAppSetup.installNpmDependencies projectDir chan)
+  let runSetupJobs =
+        ServerSetup.installNpmDependencies projectDir chan
+          `concurrently` WebAppSetup.installNpmDependencies projectDir chan
   (_, result) <- concurrently (handleJobMessages chan) runSetupJobs
   case result of
     (ExitSuccess, ExitSuccess) -> return $ Right ()
     exitCodes -> return $ Left $ setupFailedMessage exitCodes
   where
-    handleJobMessages = go Nothing (False, False)
+    handleJobMessages = runPrefixedWriter . go (False, False)
       where
-        go :: Maybe J.JobMessage -> (Bool, Bool) -> Chan J.JobMessage -> IO ()
-        go _ (True, True) _ = return ()
-        go prevJobMsg (isWebAppDone, isServerDone) chan = do
-          jobMsg <- readChan chan
+        go :: (Bool, Bool) -> Chan J.JobMessage -> PrefixedWriter ()
+        go (True, True) _ = return ()
+        go (isWebAppDone, isServerDone) chan = do
+          jobMsg <- liftIO $ readChan chan
           case J._data jobMsg of
             J.JobOutput {} ->
-              printPrefixedJobMessage prevJobMsg jobMsg
-                >> go (Just jobMsg) (isWebAppDone, isServerDone) chan
+              printPrefixedJobMessage jobMsg
+                >> go (isWebAppDone, isServerDone) chan
             J.JobExit {} -> case J._jobType jobMsg of
-              J.WebApp -> go (Just jobMsg) (True, isServerDone) chan
-              J.Server -> go (Just jobMsg) (isWebAppDone, True) chan
+              J.WebApp -> go (True, isServerDone) chan
+              J.Server -> go (isWebAppDone, True) chan
```

Comment:
Introducing state monad allowed us to hide explicit state from here!
So now here we can call prefixed writing without having to know about how it keeps its state! It just works. Before state was "previous message", now it is not visible. Super cool!

Reply:
That is very cool you abstracted away the previous message - nice! 👍🏻